src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -483,6 +483,75 @@ void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp,
   return_address_manager_->ProcessFunctionExit(event_data.tid);
 }
 
+[[nodiscard]] static std::shared_ptr<unwindstack::MapInfo> FindFileMapInfoPrecedingAnonMapInfo(
+    const std::shared_ptr<unwindstack::MapInfo>& anon_map_info) {
+  ORBIT_CHECK(anon_map_info->name().empty());
+  std::shared_ptr<unwindstack::MapInfo> map_info_it = anon_map_info->prev_map();
+  while (map_info_it != nullptr &&
+         (map_info_it->name().empty() || map_info_it->name().c_str()[0] == '[')) {
+    map_info_it = map_info_it->prev_map();
+  }
+
+  if (map_info_it == nullptr || (map_info_it->flags() & unwindstack::MAPS_FLAGS_DEVICE_MAP) != 0) {
+    // This is unexpected if anon_map_info was detected to belong to a PE.
+    return nullptr;
+  }
+  return map_info_it;
+}
+
+[[nodiscard]] static std::shared_ptr<unwindstack::MapInfo> FindFirstMapInfoForSameFile(
+    const std::shared_ptr<unwindstack::MapInfo>& file_map_info) {
+  const std::string file_path = file_map_info->name();
+  ORBIT_CHECK(!file_path.empty());
+  std::shared_ptr<unwindstack::MapInfo> first_map_info_for_file_path;
+  // Scan the maps backwards. Stop when a file mapping for a different file is found.
+  std::shared_ptr<unwindstack::MapInfo> map_info_it = file_map_info;
+  while (map_info_it != nullptr &&
+         (map_info_it->name().empty() || map_info_it->name().c_str()[0] == '[' ||
+          map_info_it->name() == file_path)) {
+    if (map_info_it->name() == file_path) {
+      first_map_info_for_file_path = map_info_it;
+    }
+    map_info_it = map_info_it->prev_map();
+  }
+  // Assigned at least by the first iteration of the loop.
+  ORBIT_CHECK(first_map_info_for_file_path != nullptr);
+  return first_map_info_for_file_path;
+}
+
+// We use PERF_RECORD_MMAP events to keep current_maps_ up to date, which is necessary for
+// unwinding.
+//
+// In addition, whenever a new executable mapping appears, it's possible that a module has been
+// newly mapped or has been re-mapped differently. We want to send a ModuleUpdateEvent in these
+// cases, so that the client has an up-to-date snapshot of the modules of the target. Ideally, for
+// each new executable file mapping we would send a ModuleUpdateEvent with the address range and
+// file for that mapping.
+//
+// But just like for orbit_object_utils::ParseMaps, things are more complicated. We observed that in
+// some cases a single loadable segment of an ELF file or a single executable section of a PE can be
+// loaded into memory with multiple adjacent file mappings. In addition, some PEs can have multiple
+// executable sections. And finally, the executable sections (and all other sections) of a PE can
+// have an offset in the file that doesn't fulfill the requirements of mmap, in which case Wine has
```

Comment:
```suggestion
// have an offset in the file that doesn't fulfill the requirements of mmap's file mapping, in which case Wine has
```

Or something along those lines.

Reply:
Good point, changed to "the requirements of mmap for file mappings".